### PR TITLE
Editor: Introduce PluginPostExcerpt slot

### PR DESCRIPTION
**Pull request by retrofox**
## What?

This PR introduces the `<PluginPostExcerpt /> `slot component to extend the Post Excerpt panel.

 ```jsx
 import { __ } from '@wordpress/i18n';
 import { PluginPostExcerpt } from '@wordpress/edit-post';
 
const MyPluginPostExcerpt = () => (
  	<PluginPostExcerpt
 		className="my-plugin-post-excerpt"
 	>
 		{ __( 'Post excerpt custom content' ) }
 	</PluginPostExcerpt>
 );
 ```

## Why?

Because I'd like to be able to extend the excerpt panel. For instance, this Jetpack plugin PR enhances the panel with AI Assistance: https://github.com/Automattic/jetpack/pull/33529

## How?

By using the [Slot&Fill core API](https://developer.wordpress.org/block-editor/reference-guides/slotfills/#how-do-they-work) \m/

## Testing Instructions

* Go to the block editor in your dev-env 
* Open the console dev tool of your client
* Confirm the Excerpt panel works as usual, as expected.
* Create a custom component to be used to fill the excerpt slot, e.g.:

```js
function MyPluginPostExcerpt() {
	return React.createElement(
		wp.editPost.__experimentalPluginPostExcerpt,
		{
			className: 'my-plugin-post-excerpt',
		},
		'🎉 Post excerpt custom content 🎇'
	)
}
```
* Register the plugin, passing the component to be rendered

```js
wp.plugins.registerPlugin( 'my-custom-excerpt-content', {
    render: MyPluginPostExcerpt,
} );
```
image

* Confirm the excerpt panel has been extended with the component defined above (MyPluginPostExcerpt):

Screenshot 2023-10-10 at 14 37 05

### CODE REVIEW

**Review by Mamaduka — COMMENTED** [review]
Thanks, @retrofox!

Do you know if this extensibility has been requested before?

The project has been strict on introducing new SlotFills in the past, and this request feels to be meeting very specific needs.

**Comment by retrofox** [comment]
Hi, there!

> Do you know if this extensibility has been requested before?

Slightly talked with @mtias about it.

File: packages/edit-post/src/components/sidebar/plugin-post-excerpt/index.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * @example
+ * ```jsx
+ * // Using ESNext syntax
+ * import { __ } from '@wordpress/i18n';
+ * import { PluginPostExcerpt } from '@wordpress/edit-post';
+ *
+ * const MyPluginPostExcerpt = () => (
+ * 	<PluginPostExcerpt
+ * 		className="my-plugin-post-excerpt"
+ * 	>
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/plugin-post-excerpt/index.js]: Nit: this is short enough to keep it on one line:

```suggestion
 * 	<PluginPostExcerpt className="my-plugin-post-excerpt">
```

File: packages/edit-post/src/components/sidebar/plugin-post-excerpt/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/**
+ * Defines as extensibility slot for the Summary panel.
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/plugin-post-excerpt/index.js]:
```suggestion
 * Defines as extensibility slot for the Excerpt panel.
```

File: packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+				<PluginPostExcerptPanel className="my-plugin-post-excerpt-custom-content">
+					Post Excerpt - Custom content
+				</PluginPostExcerptPanel>
+				<PluginPostExcerptPanel.Slot />
+			</SlotFillProvider>
+		);
+
+		expect( container ).toMatchSnapshot();
+	} );
+} );
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/index.js]: I'd say snapshot tests bring little value here. Should we add a couple of more detailed assertions here? For example:

```suggestion
/**
 * External dependencies
 */
import { render, screen } from '@testing-library/react';

/**
 * WordPress dependencies
 */
import { SlotFillProvider } from '@wordpress/components';

/**
 * Internal dependencies
 */
import PluginPostExcerptPanel from '../';

describe( 'PluginPostExcerptPanel', () => {
	test( 'renders fill properly', () => {
		render(
			<SlotFillProvider>
				<PluginPostExcerptPanel className="my-plugin-post-excerpt-custom-content">
					Post Excerpt - Custom content
				</PluginPostExcerptPanel>
				<div role="tabpanel">
					<PluginPostExcerptPanel.Slot />
				</div>
			</SlotFillProvider>
		);

		expect( screen.getByRole( 'tabpanel' ) ).toHaveTextContent(
			'Post Excerpt - Custom content'
		);
		expect(
			screen.getByText( 'Post Excerpt - Custom content' )
		).toHaveClass( 'my-plugin-post-excerpt-custom-content' );
	} );
} );
```

File: packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/__snapshots__/index.js.snap [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/__snapshots__/index.js.snap]: If you decide to abandon the snapshot testing approach, let's not forget to get rid of that file.

File: packages/edit-post/src/components/sidebar/post-excerpt/index.js [hunk 5]
```diff
@@ -13,6 +13,7 @@ import { useDispatch, useSelect } from '@wordpress/data';
  * Internal dependencies
  */
 import { store as editPostStore } from '../../../store';
+import PluginPostExcerp from '../plugin-post-excerpt';
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/post-excerpt/index.js]:
```suggestion
import PluginPostExcerpt from '../plugin-post-excerpt';
```

File: packages/edit-post/src/components/sidebar/post-excerpt/index.js [hunk 6]
```diff
@@ -44,7 +45,14 @@ export default function PostExcerpt() {
 				opened={ isOpened }
 				onToggle={ toggleExcerptPanel }
 			>
-				<PostExcerptForm />
+				<PluginPostExcerp.Slot>
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/post-excerpt/index.js]:
```suggestion
				<PluginPostExcerpt.Slot>
```

File: packages/edit-post/src/components/sidebar/post-excerpt/index.js [hunk 7]
```diff
@@ -44,7 +45,14 @@ export default function PostExcerpt() {
...
+					{ ( fills ) => (
+						<>
+							<PostExcerptForm />
+							{ fills }
+						</>
+					) }
+				</PluginPostExcerp.Slot>
```
Comment (tyxla) [inline packages/edit-post/src/components/sidebar/post-excerpt/index.js]:
```suggestion
				</PluginPostExcerpt.Slot>
```

**Review by tyxla — APPROVED** [review]
This is looking great and working well!

I've left some minor suggestions, but nothing blocking I'd say. 

Nice work 🚀

File: packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/index.js (hunk 3 above)
Comment (retrofox) [reply packages/edit-post/src/components/sidebar/plugin-post-excerpt/test/index.js]: accepting the commit and then updating the code

**Comment by retrofox** [comment]
> This is looking great and working well!
> 
> I've left some minor suggestions, but nothing blocking I'd say.

What a shame those typos. Thanks!

> Nice work 🚀

✋

**Comment by Mamaduka** [comment]
While I don't have a strong opinion for this enhancement to be merged, I think there're still some open questions - https://github.com/WordPress/gutenberg/pull/55200#pullrequestreview-1673503499.

**Comment by retrofox** [comment]
👋  

> Do you know if this extensibility has been requested before?

I don't know.

> The project has been strict on introducing new SlotFills in the past, and this request feels to be meeting very specific needs.

I think the Excerpt panel can now be extended beyond this case, which could also benefit other cases.

> While I don't have a strong opinion for this enhancement to be merged, I think there're still some open questions - [#55200 (review)](https://github.com/WordPress/gutenberg/pull/55200#pullrequestreview-1673503499).

IMHO, extending other panels such as Categories, Tags, etc should be possible, too.

I've merged the PR because I thought having a ✅ was enough. Is it ok?

**Comment by Mamaduka** [comment]
No issues with merging this specific case 👍

I wish we just had more general solution/approach for extending these panels.

**Comment by retrofox** [comment]
> I wish we just had more general solution/approach for extending these panels.

More general in which sense?
The approach keeps what we could call "the basic behavior" of the excerpt but also can extend the panel with extra functionality. It's fair to mention that I'd like to have more control over the textarea input, being able to disable it, for instance. It can be done in a follow-up.

Also, the data handling is guaranteed by using the store.

**Comment by Mamaduka** [comment]
> More general in which sense?

A general API that can enhance or replace any document panel. Currently, solutions are mixed; some panels allow overriding main components via filters (taxonomies, post thumbnail), and now this allows injecting components via SlotFills.

P.S. Here's an ongoing discussion regarding extensibility: https://github.com/WordPress/gutenberg/discussions/37448.

**Comment by retrofox** [comment]
> > More general in which sense?
> 
> A general API that can enhance or replace any document panel. Currently, solutions are mixed; some panels allow overriding main components via filters (taxonomies, post thumbnail), and now this allows injecting components via SlotFills.
> 
> P.S. Here's an ongoing discussion regarding extensibility: #37448.

I see. Thanks for sharing the issue. Very useful. I'll take a look.

**Comment by youknowriad** [comment]
I agree with @Mamaduka here, this feels a bit rushed. Also, why did we use experimental prefix, we stopped using these prefixes. It's either public or private now.

**Comment by retrofox** [comment]
# Dev Note

## Introduction of the PluginPostExcerpt Slot Component

This update introduces a new slot component within the Gutenberg editor. This enhancement allows for the extension of the Post Excerpt panel, allowing developers to customize this area for their needs.

## What's New?

The component enables the addition of custom content within the Post Excerpt panel. This is particularly useful for plugins looking to add extra functionality or information specific to their use case.

## Usage

To utilize this new feature, developers can now insert their custom content into the Post Excerpt panel by creating a component that leverages the slot. This allows for a seamless integration of custom functionalities directly within the editor.

The following example shows adding custom content to the post excerpt panel.

```jsx
import { PluginPostExcerpt } from '@wordpress/edit-post';
import { __ } from '@wordpress/i18n';

const MyPluginPostExcerpt = () => (
  <PluginPostExcerpt
    className="my-plugin-post-excerpt"
  >
    { __( 'Post excerpt custom content' ) }
  </PluginPostExcerpt>
);
```

**Comment by retrofox** [comment]
> I agree with @Mamaduka here, this feels a bit rushed. Also, why did we use experimental prefix, we stopped using these prefixes. It's either public or private now.

How should we continue to address potential issues here? [After this change](https://github.com/WordPress/gutenberg/pull/57096) to the component name, it still seems to be prefixed with `__experimental`.

**Comment by pdewouters** [comment]
> > I agree with @Mamaduka here, this feels a bit rushed. Also, why did we use experimental prefix, we stopped using these prefixes. It's either public or private now.
> 
> How should we continue to address potential issues here? [After this change](https://github.com/WordPress/gutenberg/pull/57096) to the component name, it still seems to be prefixed with `__experimental`.

This just cost me a few hours of my life that I won't get back! Finally checked the source code after trying everything else I could think of, and you have to use `import { __experimentalPluginPostExcerpt as PluginPostExcerpt } from '@wordpress/edit-post';` 

It's confusing because the docs are wrong (https://make.wordpress.org/core/2024/03/09/miscellaneous-editor-changes-in-wordpress-6-5/#introduction-of-the-pluginpostexcerpt-slot-component)  and I was working in an unfamilar codebase so wasn't sure if it was an issue with the build process as the error was very cryptic:

`react-dom.min.js?ver=18.2.0:10 Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.`

**Comment by retrofox** [comment]
Created a follow-up https://github.com/WordPress/gutenberg/pull/61188

**Comment by retrofox** [comment]
👋 @pdewouters

> This just cost me a few hours of my life that I won't get back! Finally checked the source code after trying everything else I could think of, and you have to use import { __experimentalPluginPostExcerpt as PluginPostExcerpt } from '@wordpress/edit-post';

> It's confusing because the docs are wrong (https://make.wordpress.org/core/2024/03/09/miscellaneous-editor-changes-in-wordpress-6-5/#introduction-of-the-pluginpostexcerpt-slot-component) and I was working in an unfamilar codebase so wasn't sure if it was an issue with the build process as the error was very cryptic:

I'm sorry to hear that. That isn't good at all. :-(

Additionally, #61188 has been merged, meaning the `__experimentalPluginPostExcerpt` will be deprecated in the next release. If you need to customize the excerpt part of the app, [core exposes a few resources that can help](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/#accessing-a-panel-programmatically).

**Pull request merged**